### PR TITLE
lib: feature deprecation warnings in gcc >= 4.3

### PR DESCRIPTION
**Pull request by monnerat**
Add a deprecated attribute to functions and enum values that should not be used anymore.
This uses a gcc 4.3 dialect, thus is only available for this version of gcc and newer. Note that the `_Pragma`() keyword is introduced by C99, but is available as part of the gcc dialect even when compiling in C89 mode.
    
It is still possible to disable deprecation at a calling module compile time by defining `CURL_DISABLE_DEPRECATION`.
    
Gcc type checking macros are made aware of possible deprecations.
    
Some testing support Perl programs are adapted to the extended declaration syntax.
    
Several test and unit test C programs intentionally use deprecated functions/options and are annotated to not generate a warning.

New test 1222 checks the deprecation status in doc and header files.

### CODE REVIEW

**Review by bagder — APPROVED** [review]

File: docs/libcurl/symbols-in-versions
Comment (bagder) [inline docs/libcurl/symbols-in-versions:891]: 7.87.0 and merge asap?

Comment (monnerat) [reply docs/libcurl/symbols-in-versions]: Yes, done and rebased.
Maybe wait for the CI tests to complete before merge?

Comment (bagder) [reply docs/libcurl/symbols-in-versions]: :+1:

Comment (monnerat) [reply docs/libcurl/symbols-in-versions]: CI tests complete now. Failed tests seems not related.

**Comment by bagder** [comment]
Thanks!

**Pull request closed without merging**
